tools/sof_tplg_fw_sanity_check/tplgfw_check.py

```diff
@@ -0,0 +1,739 @@
+            "id" / Int32ul,
+            "name" / String(8, "ascii"),
+            "uuid" / Bytes(16),
+            "type" / Int32ul,
+            "hash" / Bytes(32),
+            "entry_point" / Int32ul,
+            "cfg_offset" / Int16ul,
+            "cfg_count" / Int16ul,
+            "affinity_mask" / Int32ul,
+            "instance_max_count" / Int16ul,
+            "stack_size" / Int16ul,
+            "segment" / self.Module_segment[3],
+        )
+
+        self.Ext_manifest = Struct (
+            "ext_header" / self.Ext_header,
+            "padding1" / Bytes(this.ext_header.len + self.fw_header_offset - self.Ext_header.sizeof()),
+            "fw_binary_header" / self.Fw_binary_header,
+            "modules" / self.Module_entry[this.fw_binary_header.module_entries],
+        )
+
+    def parse_file(self, filepath) -> ListContainer:
+        with open(filepath, "rb") as fw :
+            return self.Ext_manifest.parse_stream(fw)
+
+if __name__ == "__main__":
+    from pathlib import Path
+
+    def parse_cmdline():
+        parser = argparse.ArgumentParser(add_help=True, formatter_class=argparse.RawTextHelpFormatter,
+            description='A Topology Reader totally Written in Python.')
+
+        parser.add_argument('--version', action='version', version='%(prog)s 1.0')
+        parser.add_argument('-t', type=str, nargs='+', required=True, help="""Load topology files for check """)
+        parser.add_argument('-f', type=str, nargs='+', required=True, help="""Load firmware files for check """)
+        parser.add_argument('-v', '--verbose', action="store_true", help="Show check state")
+
+        return parser.parse_args()
+
+
+    def main():
+        tplgFormat = TplgBinaryFormat()
+
+        cmd_args = parse_cmdline()
+
+        tplg = GroupedTplg(tplgFormat.parse_file(cmd_args.t[0]))
+
+        fw_parser = Fw_Parser()
+        fw_content = fw_parser.parse_file(cmd_args.f[0])
+        print("FW header", fw_content.ext_header, "\n")
+        if fw_content.ext_header.id == '$AE1':
+            print("IPC4 supported FW\n")
+        elif fw_content.ext_header.id == 'XMan':
+            print("IPC3 supported FW\n")
+        else:
+            print("Invalid FW")
+            return
+
+        not_found = 0
+        for widget in tplg.widget_list:
+            name = widget["widget"]["name"]
+            uuid = tplg.get_widget_uuid(widget)
+            found = False
+            if uuid is not None:
+                for m in fw_content.modules:
+                    if m.uuid == uuid:
+                        found = True
+                        if cmd_args.verbose:
+                            print("Widget: ", name, "is supported by ", m.name)
+                        break;
+
+                if found == False:
+                    not_found += 1
+                    print("Error: widget:", name, uuid, "is not supported by this FW")
+
+        if not_found == 0:
+            print("This FW can support with this topology")
```

Comment:
maybe: "FW is compatible with the topology."

please also return 0 in case of success and a negative error in case the test fails. then this script can be used as tool by other app/scripts.